Clean up QF assembly memory access

jeremylt opened the pull request:
Cleaning up this memory lets us more fully mirror the dual memory spaces on the GPU by making writable array pointers from `CeedVectorGetArray[Write]` invalid after calling `CeedVectorRestoreArray[Write]`

Fixes #1454

jeremylt commented on backends/blocked/ceed-blocked-operator.c:
```diff
@@ -471,52 +470,42 @@ static inline int CeedOperatorLinearAssembleQFunctionCore_Blocked(CeedOperator o
-        }
-        num_active_in += size;
-        CeedCallBackend(CeedVectorRestoreArray(impl->q_vecs_in[i], &q_vec_array));
+        num_active_in += field_size;
```
Yeah, it's the total number of input/output... modes? active_size_in maybe?  I disliked the wording I originally picked years ago and still don't like it now.

jrwrigh replied:
I like `active_size_in`. Or `total_active_comps_in`, but I'm afraid that might be confusing with the restrictions number of components which may not match up (e.g. in the case of ` CEED_EVAL_GRAD` on a field.

jeremylt replied:
Yeah, I don't think we can use the term 'comp' and we already use the phrase 'size' when building QF fields so that's starting to feel more right

jeremylt commented:
Ok, a bit of a fuller write-up.

There are two fixes here in this PR

1) CPU backends previously "cheated" when assembling QFunctions. We'd `CeedVectorGetArray` a pointer to the CeedVector q-vec inputs and outputs and set portions of those corresponding to each part of the input/output as the array for separate CeedVector. For example, an input with 2 components with CEED_EVAL_GRAD in 3d would be wrapped by 6 CeedVector. But this violates the access assumptions - we should not use an array pointer after `CeedVectorRestoreArray`. I dropped this and just directly accessed and updated the array of interest as we assembled the QF.

2) With this fixed, all CPU backends internally follow the memory access assumptions. I could thus update the memcheck backends to verify these assumptions are being followed for write access. Now the memcheck backends create a writable buffer that is given to the users when `CeedVectorGetArray[Write]` is called. The changes in this buffer are only applied to the "true" data array once `CeedVectorRestoreArray[Write]` is called. Also, this writable buffer is invalid once `CeedVectorRestoreArray[Write]` is called, so my hack of using a copy of the ref to the returned pointer would segfault. Now the memcheck backend is pretty darn close to mirroring the double memory space model in GPU backends.
